pyo3-derive-backend/src/pyproto.rs

```diff
@@ -129,42 +129,46 @@ fn inventory_submission(py_methods: Vec<TokenStream>, ty: &syn::Type) -> TokenSt
     }
 }
 
-fn slot_initialization(
+fn submit_protocol_methods(
     method_names: HashSet<String>,
     ty: &syn::Type,
     proto: &defs::Proto,
 ) -> syn::Result<TokenStream> {
-    // Collect initializers
-    let mut initializers: Vec<TokenStream> = vec![];
-    for setter in proto.setters(method_names) {
-        // Add slot methods to PyProtoRegistry
-        let set = syn::Ident::new(setter, Span::call_site());
-        initializers.push(quote! { table.#set::<#ty>(); });
+    if proto.extension_trait == "" {
+        return Ok(quote! {});
+    }
+    let ext_trait: syn::Path = syn::parse_str(proto.extension_trait)?;
+    let mut tokens: Vec<TokenStream> = vec![];
+    if proto.name == "Buffer" {
+        // For buffer, we construct `PyProtoMethods` from PyBufferProcs
+        tokens.push(quote! {
+            let mut proto_methods = pyo3::ffi::PyBufferProcs::default();
+        });
+        for getter in proto.slot_getters(method_names) {
+            let get = syn::Ident::new(getter, Span::call_site());
+            let field = syn::Ident::new(&format!("bf_{}", &getter[4..]), Span::call_site());
+            tokens.push(quote! { proto_methods.#field = Some(<#ty as #ext_trait>::#get()); });
+        }
+    } else {
+        // For other protocols, we construct `PyProtoMethods` from Vec<ffi::PyType_Slot>
+        tokens.push(quote! { let mut proto_methods = vec![]; });
+        for getter in proto.slot_getters(method_names) {
+            let get = syn::Ident::new(getter, Span::call_site());
+            tokens.push(quote! { proto_methods.push(<#ty as #ext_trait>::#get()); });
```

Comment:
Maybe we can do the creation of `PyType_Slot` objects and conversion to `*mut void` here?

Then the slot getters can return well-typed function pointers.

```suggestion
            tokens.push(quote! { proto_methods.push(pyo3::ffi::PyType_Slot {
                slot: #slot,
                pfunc: <#ty as #ext_trait>::#get()
            }); });
```

We would need to move the `#slot` definitions into the proc macro. I think that could be a good idea anyway to help keep the definitions in one place.

Reply:
Sounds nice :+1: !